File: Python/trie-based_search.py
class Node:
    def __init__(self):
        self.children = {}
        self.val = None
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word):
        curr = self.root
        for char in word:
            if char in curr.children:
                curr = curr.children[char]
            else:
                new_node = Node()
                curr.children[char] = new_node
                curr = new_node
        curr.val = word
        curr.is_end_of_word = True

    def search_word(self, curr):
        is_end = False

        while is_end is False:
            if curr.is_end_of_word:
                is_end = True
            else:
                curr = next(iter(curr.children.values()))
        return curr.val

    def search_node(self, word):
        curr = self.root

        for char in word:
            if char in curr.children:
                curr = curr.children[char]
            else:
                return None
        return curr
    
    def delete(self, word):
        curr = self.root
        path = []

        for char in word:
            if char not in curr.children:
                return False
            path.append((curr, char))
            curr = curr.children[char]
        
        if not curr.is_end_of_word:
            return False
        
        curr.is_end_of_word = False
        curr.val = None

        # Remove associated non-words
        for par, char in reversed(path):
            child = par.children[char]

            if child.is_end_of_word == true or len(child.children) > 0:
                break
            del par.children[char]
        return True

    """
    Generate the display_trie function that uses the terminal and print statements to properly display the trie based on this code [Code with a blank display_trie function here]

    Used: Sonnet 5 (High Effort)
    """
    def display_trie(self):
        """Print the trie as an indented tree in the terminal.
 
        Each edge label is the character stored on that link. A node
        that completes a word is marked with '*' followed by the word
        itself, so full words are easy to spot even when they share a
        path with longer words (e.g. "car" inside "carrot").
        """
        print(".")
        self._display(self.root, prefix="")
        print("\n")
 
    def _display(self, node, prefix):
        chars = sorted(node.children.keys())
        for i, char in enumerate(chars):
            child = node.children[char]
            is_last = (i == len(chars) - 1)
            connector = "└── " if is_last else "├── "
 
            label = char
            if child.is_end_of_word:
                label += f"  *[{child.val}]"
 
            print(prefix + connector + label)
 
            extension = "    " if is_last else "│   "
            self._display(child, prefix + extension)


def main():
    trie = Trie()
    words = ["bar", "bat", "bike", "cat", "car", "carrot", "dog", "dot", "free", "freedom"]

    for word in words:
        trie.insert(word)

    trie.display_trie()



main()

File: C/trie-based_search.c
/*
*   Structures
*/

// Node structure for trie
struct HashMap;

struct Trie_Node {
    struct HashMap *hashmap;
    char *val;
    char character;
    bool is_end_of_word;
};


// Hash map structure
struct HashMap {
    struct Trie_Node **hash_buckets; // A pointer to an array of child nodes
    int size; // Number of buckets in the hash map
};



/*
*    Hash
*/

// Create an empty hash map with a given size
struct HashMap* create_hash_map() {
    struct HashMap* map = malloc(sizeof(struct HashMap));

    map->size = 26;
    map->hash_buckets = malloc(map->size * sizeof(struct Trie_Node*));

    for (int i = 0; i < map->size; i++) {
        map->hash_buckets[i] = NULL; // Initialize each bucket to null character
    }

    // Return the pointer to the created hash map
    return map;
};

// Computes and returns the index for a given value in the HashMap
int hash_map_function(char val) {
    return (unsigned char)(val-'a') % 26;
}


// Inserts a value into the HashMap
void insert_to_hash_map(struct HashMap* map, struct Trie_Node *node) {
    int index = hash_map_function(node->character);

    if (index < 0 || index >= map->size) return; // Invalid index check

    map->hash_buckets[index] = node;
}


// Free all memory allocated for the hash map and its contents
void free_hash_map_and_contents(struct HashMap* map) {
    // Empty map safty check
    if (!map) return;

    free(map->hash_buckets);
    free(map);
}



/*
*    Trie
*/

// Function for inserting into the trie
void insert_node(struct Trie_Node *root, char *word) {
    struct Trie_Node *curr_node = root;
    for (int i=0; word[i] != '\0'; i++) {
        char c = word[i];
        if (curr_node->hashmap->hash_buckets[hash_map_function(c)] != NULL) {
            curr_node = curr_node->hashmap->hash_buckets[hash_map_function(c)];
        } else {
            struct Trie_Node *new_node = malloc(sizeof(struct Trie_Node));
            new_node->hashmap = create_hash_map();
            new_node->val = NULL;
            new_node->character = c;
            new_node->is_end_of_word = false;
            insert_to_hash_map(curr_node->hashmap, new_node);
            curr_node = new_node;
        }
    }

    curr_node->is_end_of_word = true;
    curr_node->val = word;
}


// Function to search for word
char* search_word(struct Trie_Node *root, char *needle) {
    struct Trie_Node *curr_node = root;

    if (needle == NULL || needle[0] == '\0') {
        return NULL;
    }

    for (int i=0; needle[i] != '\0'; i++) {
        char c = needle[i];
        if (curr_node->hashmap->hash_buckets[hash_map_function(c)] == NULL) {
            if (curr_node->is_end_of_word == true) {
                return curr_node->val;
            } else {
                return NULL;
            }
        } else {
            curr_node = curr_node->hashmap->hash_buckets[hash_map_function(c)];
        }
    }

    while (curr_node->is_end_of_word == false) {
        Struct Trie_Node *temp_node = NULL;
        for (int i=0; i<curr_node->hashmap->size) {
            if (curr_node->hashmap->hash_buckets[i] != NULL) {
                temp_node = curr_node->hashmap->hash_buckets[i];
                break;
            }
        }

        if (temp_node == NULL) {
            return NULL;
        } else {
            curr_node = temp_node;
        }
    }
    return curr_node->val;
}


// Function to search for a specific node
struct Trie_Node *search_node(struct Trie_Node *root, char *word) {
    struct Trie_Node *curr_node = root;
    bool is_end = false;
    for (int i=0; word[i] != '\0'; i++) {
        char c = word[i];
        if (curr_node->hashmap->hash_buckets[hash_map_function(c)] != NULL) {
            curr_node = curr_node->hashmap->hash_buckets[hash_map_function(c)];
        } else {
            return NULL;
        }
    }
    return curr_node;
}


// Function to delete node from trie
bool delete_node(struct Trie_Node *root, char *word) {

}


// Function to delete trie
void delete_trie(struct Trie_Node *root) {

}


// Function to display trie
void display_trie(struct Trie_Node *root) {

}


// Driver code
int main() {
    struct Trie_Node *trie = malloc(sizeof(struct Trie_Node));
    trie->hashmap = create_hash_map();
    trie->val = NULL;
    trie->character = '\0';
    trie->is_end_of_word = false;

    insert_node(trie, "bar");
    insert_node(trie, "bat");
    insert_node(trie, "bike");
    insert_node(trie, "cat");
    insert_node(trie, "car");
    insert_node(trie, "carrot");
    insert_node(trie, "dog");
    insert_node(trie, "dot");
    insert_node(trie, "free");
    insert_node(trie, "freedom");

    display_trie(trie);

    delete_trie(trie);
}